### PR TITLE
Remove vendored modules and use bundler with ESM imports

### PR DESCRIPTION
**Pull request by manzt**
@kevinyamauchi pointed me to https://github.com/napari/napari/pull/1909, really awesome @pwinston!

I know the JS is proof-of-concept, but this PR adds a more modern JS dev environment and removes the vendor-ed JS libs. Rather than relying on the order of imports in `templates/viewer.html` and various global functions, each library is imported as an ESM module and bundled by [`esbuild`](https://esbuild.github.io/) into `/static/bundle.js`.

To bundle the JS source, you need to have `npm` installed:

```bash
$ cd js
$ npm install && npm run build # creates /static/bundle.js
```

Esbuild is written in GO and bundling is nearly instant (esbuild can bundle the THREE.js source 10x in 0.37s). Pairing the build command with something like [`watchexec`](https://github.com/watchexec/watchexec) means you can get live reloading while making changes.

```bash
$ watchexec --ext js npm run build # watches for changes to js source and will re-bundle
```

This could be an overload of tooling (and feel free to ignore these suggestions), but using tools from the js ecosystem means that adding or changing dependencies is much easier in the future. Cheers!

### CODE REVIEW

File: js/src/utils.js [hunk 1]
```diff
@@ -5,10 +5,14 @@
...
 // We define or ortho camera to view the space [0..1], but then we zoom
 // a bit so that area has a border around it.
-ZOOM = 0.8;
+import * as THREE from 'three';
+import { TrackballControls } from 'three/examples/jsm/controls/TrackballControls';
+import io from 'socket.io-client';
 
-var externalParams;
-function defineExternalParams() {
+const ZOOM = 0.8;
```
Comment (manzt) [inline js/src/utils.js:12]: Modules require variable declarations.

**Comment by pwinston** [comment]
Thanks this looks like a good step forward. I appreciate it!

I'm not that used to modern Javascript, but my impression is build steps are very common now days. I have used some JS projects that had them. So this seems like a good thing to try.

So much of webmon today is verbatim from the SocketIO/WebGL sample I started with:
https://github.com/ageller/FlaskTest

While that sample was nice in that it did exactly what I wanted. Combining Flask-SocketIO with three.js/WebGL. I think both the Python side (now `webmon.py`) and the Javascript side were just minimal examples and not something to really build on. So both probably do need a lot of iteration.

I try this out and merge it if it's working for me.

**Comment by manzt** [comment]
> I'm not that used to modern Javascript, but my impression is build steps are very common now days. I have used some JS projects that had them. So this seems like a good thing to try.

I think the benefit of the build step is that you can easily try and upgrade modules that are distributed via npm (no copying around JS files) and tree-shaking (dead code elimination) of those modules. The main headache is that most bundlers are written in JS and the build step can be slow for large projects, making development somewhat painful.

> I'll try this out and merge it if it's working for me.

Let me know if you have any questions!

**Comment by pwinston** [comment]
Looks great, every helpeful. I suspect we'll have lots of new dependencies on the JS side soon. Probably some 2D charts/graphs soon, any suggestions? Seems like there are lots to choose from.

**Comment by manzt** [comment]
> Any suggestions?  Seems like there are lots to choose from.

Yes, _too many_ to choose from. I'd suggest looking into [Vega-Lite](https://vega.github.io/vega-lite/) for a start. It's a "high-level grammar of interactive graphics" that's built on top of D3 and expressed in JSON. It offers a powerful set of primitives to compose fairly sophisticated interactive graphics with very little effort (especially with the [JS api](https://observablehq.com/@vega/vega-lite-api)). There are also _tons_ of examples on Observable HQ:

- Introduction to Vega-Lite: https://observablehq.com/@uwdata/introduction-to-vega-lite
- Vega-Lite with live streaming updates: https://observablehq.com/@vega/vega-lite-with-streaming-data-updates (this one likely most relevant to `webmon`)
- Apache Arrow in Vega-Lite: https://observablehq.com/@vega/apache-arrow-in-vega-lite (if you want to quickly serialize a pandas dataframe)

PS: You can actually write Vega-Lite in Python with [Altair](https://github.com/altair-viz/altair) and export the validated JSON to your web-app.

**Comment by pwinston** [comment]
Thanks. Looks neat. I'll check it out. The other thing I want to get in place early is some sort of tabs or menu of pages. 

So today localhost:5000/viewer is this full screen three.js view of the tiles. I'm thinking that would be just one of N pages. And there is some always visible list of such pages as tabs or maybe on the left side gives us more room. 

For example the tiles view is one page. But then maybe another is page has ChunkLoader graphs and tables. Another is something not async/octree related at tree at all. 

Basically so webmon has some overall structure and look that people can add any number of "pages" to. But navigate between then small easily. And so we avoid the temptation to cram everything on to one giant page. 

I've setup such things before maybe with just bootstrap and HTML. But again looking for suggestions. Something modern and simple and functional.

If webmon gets used we can always redo the overall look. So this doesn't have to last for the ages. Just something to start with. Any ideas? Or do you even think that's even a good idea?

Sorry to hijack this PR for discussion.

**Comment by manzt** [comment]
It makes total sense to me to break into separate pages, and not one giant SPA.

> If webmon gets used we can always redo the overall look. So this doesn't have to last for the ages. Just something to start with. Any ideas? Or do you even think that's even a good idea?

From what you've described so far, I think Bootstrap + HTML is a probably a good starting place. You can use Flask's templates to create each route (e.g. `viewer.html`) and have each route derive from a base template that has a navbar to link to each separate page (https://stackoverflow.com/questions/32567100/add-a-nav-bar-to-all-templates). 

### Templates
```html
// viewer.html 
{% extends 'base.html' %}

{% block content %}
  <div id="WebGLContainer">
{% endblock %}
{% block scripts %}
  <script type="module" scr="/static/viewer.js"></script>
{% endblock scripts %}
```

```html
// other-page.html 
{% extends 'base.html' %}

{% block content %}
  <div id="WebGLContainer">
{% endblock %}
{% block scripts %}
  <script type="module" scr="/static/other-page.js"></script>
{% endblock scripts %}
```

### Bundling JS for multiple routes

`esbuild` supports code splitting for multiple entrypoints, which means you can write all your JS in one place, and the final static assets are the minimal JS needed for each route. (e.g. shared function/variables are imported from `/static/chunk.PGDAFG3D.js` into `/static/viewer.js` and `/static/other-page.js`), so you'll want to write a unique `/src/<route>.js` for each `/template/<route>.html`.

```bash
$ esbuild --bundle --splitting --minify --format=esm --outdir=../static/ src/viewer.js src/other-page.js 
$ ls ../static/
chunk.PGDAFG3D.js viewer.js other-page.js 
```

> NOTE: You'll need to include `type="module"` in the script tag on each route template (like above) and add the `--splitting --format=esm` to esbuild.


### Tailwind CSS

Bootstrap is totally fine, but I might suggest looking into [Tailwind CSS](https://tailwindcss.com/). It's becoming very popular and is more modern, and a bit sleeker IMO. You can still template using HTML and then use Tailwind like you would Bootstrap (https://tailwindcss.com/docs/installation#using-tailwind-via-cdn), or you can have a `tailwind.config.js` file and have a postCSS step to have a custom theme and purge unused CSS classes.

### A note on JS frameworks (and frontend UI libraries)

I avoided suggesting a JS framework simply because I think it would add an extra level of complexity that probably isn't necessary for the time being. The proposal above fits into the Flask way of creating a web-app (static assets + dynamically rendered templates), and I'd like to avoid making you "buy-in" to frontend framework if at all possible.

With that said, if the UI starts to become sufficiently complex, you'll probably want to reach for something like React, Vue, or Svelte for writing the UI. For each of these, there are frameworks like [`Next.js`](https://nextjs.org/) (React), [`Nuxt.js`](https://nuxtjs.org/) (Vue), and [`Sapper`](https://sapper.svelte.dev/) (Svelte) which provide an all-in-one solution to build sites. In this case, your "build" step would create a full static export of the app (folder of all assets, js, html, and css) Flask would just serve the contents of that folder and handle socket.io. It would no longer handle templating and rendering routes dynamically since the frontend framework has build all routes and endpoints.

> Sorry to hijack this PR for discussion.

No problem, happy to bounce ideas. The web moves too fast.

**Pull request merged**